Fix index messages

lingshanng opened the pull request:
fix https://github.com/AY2122S1-CS2103T-F13-3/tp/issues/125
fix https://github.com/AY2122S1-CS2103T-F13-3/tp/issues/241

Currently, when index is not nonZeroUnsignedInteger, it simply shows invalid command format, instead of invalid index message.
With this fix, invalid index message will show when 
1. Preamble split by whitespace has the same number of args as number of required indices.
2. All required arguments are present and valid
2. The index parsed is not nonZeroUnsignedInteger or is out of range

(For reference) previous implementation suggestion that had loophole: https://github.com/AY2122S1-CS2103T-F13-3/tp/pull/118#discussion_r734926352

## Example of messages shown

**Valid commands:**
- `edit 1 n/alex`
- `ledit   1      1    n/alex` (with many white space between the indices)

**Commands that show `MESSAGE_INVALID_COMMAND_FORMAT`**:
- `edit`
- `edit n/alex` (empty preamble)
- `edit 1 1 n/alex` (extra character(s) in preamble)
- `edit 1 nn/alex a/abc` (correct index with invalid prefix)
- `ledit 1 subject/math` (missing index) 
- `paid 1 0`, `ladd 0` (missing field have priority over invalid index)

**Commands that show `You must provide at least one field to edit!` + `MESSAGE_USAGE`**:
- `edit 0`
- `ledit 0 0`

**Commands that show `The student index provided is invalid`**:
- `edit 0 n/alex`
- `edit a n/alex`
- `edit 10000000000 n/alex`  
- `edit +1 n/alex` 
- `edit 10 n/alex` (out of student range)
- `paid 0 0 amt/100` (student index error have priority over lesson index error)


**Commands that show `The lesson index provided is invalid`**:
- `paid 1 abc amt/100`
- `paid 1 10 amt/100` (out of lesson range)
- `ldelete 1 0`

**Field errors have priority over index errors to be consistent with out of range error's behaviour.**
**Commands that show `Field.MESSAGE_CONSTRAINTS`**:
- `edit 0 n/alex!?`
- `edit 10 n/alex!?` (out of range)
- `ledit 0 1 subject/ma!!`
- `ladd 0 date/Oct time/1430-1600 rates/37.50 subject/Science`  (invalid date)
- `paid 1 0 amt/abc`

eeliana commented:
If there are multiple whitespaces in preamble like in case 2 of MESSAGE_INVALID_COMMAND_FORMAT, maybe we want do help replace them? It could be hard for users to see amidst all the other words in the command.

https://stackoverflow.com/questions/2932392/java-how-to-replace-2-or-more-spaces-with-single-space-in-string-and-delete-lead

eeliana commented on src/main/java/seedu/address/logic/parser/ParserUtil.java:
```diff
@@ -36,9 +36,37 @@
+     * preamble string is split by whitespace into an array of strings.
+     *
+     * @param preamble
+     * @return
```
Missing

eeliana reviewed (approved):
LGTM overall.

Xiaoyunnn reviewed (commented):
valid student index and invalid lesson index within range:
![image](https://user-images.githubusercontent.com/76084739/139629742-ab65ff22-029d-494a-919e-dcad731fc46b.png)

valid student index and invalid lesson index beyond range:
![image](https://user-images.githubusercontent.com/76084739/139629768-65a67778-e0ef-46ef-98a8-e9494bc88189.png)

invalid student index within range:
![image](https://user-images.githubusercontent.com/76084739/139630043-b5d8329c-18c3-4d09-8faf-a85e14d36a0e.png)

invalid student index beyond range:
![image](https://user-images.githubusercontent.com/76084739/139630061-ec2b8a14-c381-448a-99e6-9123e414ba6b.png)

should we specify for lesson commands if its the student or lesson index that is invalid?

eeliana commented on src/main/java/seedu/address/logic/parser/LessonEditCommandParser.java:
```diff
@@ -41,18 +44,12 @@ public LessonEditCommand parse(String args) throws ParseException {
-        assert indices.length == 2;
-        Index studentIndex = indices[0];
-        Index lessonIndex = indices[1];
+        Index studentIndex = ParserUtil.parseIndex(preamble[STUDENT_INDEX_ZERO_BASED]);
+        Index lessonIndex = ParserUtil.parseIndex(preamble[LESSON_INDEX_ZERO_BASED]);
```
I don't think it's a big issue but perhaps you can do the catching and throwing of relevant error messages here like AB3 did since the 2 indices are separate here (same for ldelete?)

lingshanng commented:
I've updated it to specify whether it's student or lesson index that is invalid. 
There's a few other changes which I've updated in this PR's description.

Xiaoyunnn reviewed (approved):
LGTM. Just that these 2 messages has no punctuation at the end unlike other error messages
![image](https://user-images.githubusercontent.com/76084739/139849042-2c9334a5-c249-4ebc-81c3-024a9c28cfd2.png)
![image](https://user-images.githubusercontent.com/76084739/139849063-1fdf37cd-987f-46a4-a9f0-c0eefbd93541.png)
